### DISCUSSION TITLE
Hope to know more about the algorithm behind the hyperbola fitting

### DISCUSSION THREAD

**Discussion by shuolingliu-git** [emanuelhuber/RGPR#48] [Show and tell]
Dear Mr. Emanuel Huber, 

Thanks so much for creating this! It helps me a lot. 

I desperately want to know the math / algorithm behind the hyperbola fitting. Cuz I've been doing research on hyperbolic fitting. Is it the hough transform, neural networks or algorithm based on orthogonal distance? Could you please explain me the details of how this function( hyperbolaFit() ) is implemented? Or could you please tell me what books or literature referred to this function?

Hoping to hear from you soon!
Best wishes!

Sincerely,
hhhhappyyyy

**Comment by emanuelhuber** [comment]
Hello hhhhappyyyy

Thank you for contacting me. I moved the "issue" you opened into "discussion" because it is not really an issue about RGPR.

The function `hyperbolaFit()` estimate the hyperbola parameters given some points `(x, y)` that the user selected (e.g., by extracting some point coordinates a hyperbola). This is a manual hyperbola fitting and not an automatic one. The fitting of the hyperbola to the points uses a simple linear regression (works enough well).

Automatic hyperbola fitting is not yet implemented though I am currently testing an approach based on the so-called probabilistic hough transform (with selection of the region of interest using the local trace extremum instead of using an edge detection algorithm).
Could this approach interest you? What do you aim to do? Is it for research purposes or "just" for work? 

Let me know.
Best,
Emanuel

**Comment by shuolingliu-git** [comment]
Dear Mr. Emanuel Huber,

Thank you so much for your prompt response.

I'm really interested in this topic. I am trying to use GPR explore the spatial position and the structure of the tree roots. I want to do hyperbola fitting for root radius and position estimation. In fact I am currently testing an approach based on the random hough transform with Matlab. But it is not yet implemented because of some errors I got. I'm still trying to figure it out.

I have a question. If there are several hyperbolas in an image, whether the function hyperbolaFit() can only fit one hyperbola first, then the second, until all the curves are fitted ?

Best wishes!
hhhhappyyyy

**Reply by shuolingliu-git** [reply]
Additionally, I do it for research purposes,so I'm interested in the details of function implementation.

**Reply by emanuelhuber** [reply]
OK! I will send you my current working code by Friday. Let me know if you need any collaboration or support.
Best,
Emanuel

**Reply by shuolingliu-git** [reply]
Dear Mr. Emanuel Huber,

Thank you very much for your time and your generous offer to help. 
Have a good day!

best,
hhhhappyyyy

**Comment by emanuelhuber** [comment]
Hello hhhhappyyyy

I have a bug in my script and I will first fix it before I send it to you.
Nevertheless, I can explain you the approach I am testing with the randomized/probabilistic Hough transform (although it does no more look like a Hough transform).
The principle of the Hough transform: select 3 pixels somewhere on your radargram, fit an hyperbola to this 3 points, extract the amplitudes along the hyperbola and evaluate if this could be a real hyperbola

1. Reduce the region of interest (ROI): I extract the local extremum (min/man) on each trace. 
2. I sample 3 pixel from the extracted local extremum, with coordinates `xv, xy`
3. I fit an hyperbola to the coordinates of these 3 pixels:
    ```
    hyp <- RGPR::hyperbolaFit(xy, yv)
    ```
4. I extract the coordinates of the pixels along the hyperbola
5. Knowing the pixel coordinates I can extract the amplitudes of these pixels
6. Then I can compute some statistics (e.g., sum(ampl))
7. Then I set a threshold to keep only the hyperbola that may correspond to true hyperbolas, e.g., for which sum(amp) > threshold
8. I did not implement step 8. Here the idea is to use the selected hyperbola to define region of interest in which a finer fitting of the hyperbola can be done.

Hope it helps a little be.
Thank you for your patience.
Best,

Emanuel

**Reply by shuolingliu-git** [reply]
Dear Mr. Emanuel Huber,

Many thanks for your effort! I think your instruction is really helpful. I will test it in my own program.

Have an awesome weekend.

Best wish,
hhhhappyyyy
